services/mailer/incoming/incoming.go

```diff
@@ -0,0 +1,422 @@
+		return true
+	}
+	autoReply := h.Get("X-Autoreply")
+	if autoReply == "yes" {
+		return true
+	}
+	autoRespond := h.Get("X-Autorespond")
+	return autoRespond != ""
+}
+
+// searchTokenInHeaders looks for the token in To, Delivered-To and References
+func searchTokenInHeaders(h mail.Header) string {
+	if addressTokenRegex != nil {
+		to, _ := h.AddressList("To")
+		deliveredTo, _ := h.AddressList("Delivered-To")
+		for _, list := range [][]*net_mail.Address{
+			to,
+			deliveredTo,
+		} {
+			for _, address := range list {
+				match := addressTokenRegex.FindStringSubmatch(address.Address)
+				if len(match) != 2 {
+					continue
+				}
+
+				return match[1]
+			}
+		}
+	}
+
+	references := h.Get("References")
+	for {
+		begin := strings.IndexByte(references, '<')
+		if begin == -1 {
+			break
+		}
+		begin++
+
+		end := strings.IndexByte(references, '>')
+		if end == -1 || begin > end {
+			break
+		}
+
+		match := referenceTokenRegex.FindStringSubmatch(references[begin:end])
+		if len(match) == 2 {
+			return match[1]
+		}
+
+		references = references[end+1:]
+	}
+
+	return ""
+}
+
+type MailContent struct {
+	Content     string
+	Attachments []*Attachment
+}
+
+type Attachment struct {
+	Name    string
+	Content bytes.Buffer
+}
+
+// getContentFromMailReader reads the plain content and the attachments from the mail.
+// If there is only HTML content, it gets converted to plain text.
+// A potential reply/signature gets stripped from the content.
+func getContentFromMailReader(mr *mail.Reader) (*MailContent, error) {
+	contentText := ""
+	contentHTML := ""
+	attachments := make([]*Attachment, 0, 1)
+
+	for {
+		p, err := mr.NextPart()
+		if errors.Is(err, io.EOF) {
+			break
+		} else if err != nil {
+			return nil, fmt.Errorf("next part failed: %w", err)
+		}
+
+		switch h := p.Header.(type) {
+		case *mail.InlineHeader:
+			contentType, _, err := h.ContentType()
+			if err != nil {
+				return nil, fmt.Errorf("ContentType failed: %w", err)
+			}
+
+			if contentType == "text/plain" {
+				if contentText != "" {
+					continue
+				}
+			} else if contentType == "text/html" {
+				if contentHTML != "" {
+					continue
+				}
+			} else {
+				continue
+			}
+
+			data, err := io.ReadAll(p.Body)
+			if err != nil {
+				return nil, fmt.Errorf("read body failed: %w", err)
+			}
+
+			switch contentType {
+			case "text/plain":
+				contentText = string(data)
+			case "text/html":
+				contentHTML = string(data)
+			}
+		case *mail.AttachmentHeader:
+			attachment := &Attachment{}
+			attachment.Name, _ = h.Filename()
+			_, err := io.Copy(&attachment.Content, p.Body)
+			if err != nil {
+				return nil, fmt.Errorf("read attachment failed: %w", err)
+			}
+
+			attachments = append(attachments, attachment)
+		}
+	}
+
+	if contentText == "" && contentHTML != "" {
+		var err error
+		contentText, err = html2text.FromString(contentHTML)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return &MailContent{
```

Comment:
The contentText maybe not UTF8 in some countries. Maybe we need to detect and convert.

Reply:
I use a different library now which does handle other encodings.